### PR TITLE
Check index for built distributions

### PR DESCRIPTION
**Pull request by jakirkham**
Hopefully 🍀 ...

Fixes https://github.com/conda-forge/conda-smithy/issues/270
Fixes https://github.com/Anaconda-Platform/support/issues/60

As we were requesting a list of all files (easily in the 1000s) from the `conda-forge` channel, this was likely slamming the API to process all of this. As this takes a fair bit of time, it was resulting in request timeouts. Thus when the request timed out this caused uploads to fail.

This borrows from @mwcraig's PR ( https://github.com/SciTools/conda-build-all/pull/28 ) to switch to using the index. Unlike asking for the file list from the API, asking for the index should be a near immediate operation. This is because anaconda.org and conda were designed cleverly to ensure the index exists as a separate file from the packages. As a result, we should be able to just download this JSON file (the index) and check against that instead.

Would really appreciate feedback from @bollwyvl and @dsludwig to see if this will solve our issues.

### CODE REVIEW

**Comment by jakirkham** [comment]
Thanks for asking this functionality in `conda-build-all`, @ericdill. Seems like it is still paying dividends 💰.

**Comment by bollwyvl** [comment]
Seems like a good stopgap approach to me... not too long ago we re-jiggered the server-side cache regime on `repodata.json*` which should help ensure that this is hitting a cached copy unless the label _actually did change_ within the last five minutes. However, it's still a big file (3mb) to get for basically a single bit of information. Hopefully we can get you that bit in production soon with https://github.com/Anaconda-Platform/support/issues/60 ... we got to beta yesterday, still shaking things down before turning on prod...

**Comment by jakirkham** [comment]
Agreed being able to query by filename is going to be crucial. Though you are absolutely right that downloading the index is not a great thing. We are planning to break this script out of `conda-smithy` so we can iterate on it faster without needing multiple re-renderings. See PR ( https://github.com/conda-forge/conda-forge-build-setup-feedstock/pull/25 )

**Comment by pelson** [comment]
👍  on this change. I'm not eager to do it in the feedstock though - would you be OK pulling it out into conda-forge-build-setup and calling it from the feedstock? The conda-smithy change is then to remove the upload script from the repo, and remove it from a feedstock at rerender time.

**Comment by pelson** [comment]
Should read my notification first (all 2000 of them 😄 ). https://github.com/conda-forge/conda-forge-build-setup-feedstock/pull/25 puts this script into conda-forge-build-setup.

**Comment by jakirkham** [comment]
Sorry, it's probably mainly me pinging you. 😜

Still need to move this change over there too. See PR ( https://github.com/conda-forge/conda-forge-build-setup-feedstock/pull/26 ). Literally a copy-pasta of this script there.

FYI am going to be out today and probably won't be back until Tuesday. So if there is something urgent, feel free to steer the ship as you see fit, @pelson. 🚢 Otherwise will fix the re-rendering when I'm back if not sooner.

File: conda_smithy/feedstock_content/ci_support/upload_or_check_non_existence.py [hunk 1]
```diff
@@ -57,17 +58,23 @@ def upload(cli, meta, owner, channels):
         os.remove('binstar.token')
 
 
-def distribution_exists_on_channel(binstar_cli, meta, owner, channel='main'):
+def distribution_exists_on_channel(binstar_cli, owner, metadata, channel='main'):
```
Comment (jakirkham) [inline conda_smithy/feedstock_content/ci_support/upload_or_check_non_existence.py]: This is breaking the API elsewhere and doesn't match what we are doing in other places in this code. So am going to adjust the code to match.

**Comment by jakirkham** [comment]
So PR ( https://github.com/conda-forge/conda-smithy/pull/301 ) will actually make use of this external upload script. The sooner we can get that in the better.

**Pull request closed without merging**